File: riveredge-backend/src/apps/kuaizhizao/schemas/purchase_arrival.py
"""采购到货延期填报 Schema"""

from datetime import date, datetime
from typing import List, Optional

from pydantic import BaseModel, Field, ConfigDict

from core.schemas.base import BaseSchema


DELAY_REASON_OPTIONS = (
    "supplier_capacity",
    "logistics",
    "quality_rework",
    "drawing_change",
    "payment_delay",
    "other",
)


class PurchaseArrivalDelayReportCreate(BaseSchema):
    purchase_order_item_id: int = Field(..., description="采购订单明细ID")
    delay_reason: str = Field(..., max_length=50)
    estimated_arrival_date: date = Field(..., description="预计新到货日")
    impact_description: Optional[str] = Field(None, description="影响说明")
    attachments: Optional[List[dict]] = None
    notes: Optional[str] = None


class PurchaseArrivalDelayReportUpdate(BaseSchema):
    delay_reason: Optional[str] = Field(None, max_length=50)
    estimated_arrival_date: Optional[date] = None
    impact_description: Optional[str] = None
    attachments: Optional[List[dict]] = None
    notes: Optional[str] = None


class ApproveDelayReportRequest(BaseModel):
    approved: bool = True
    review_remarks: Optional[str] = None


class PurchaseArrivalDelayReportResponse(BaseSchema):
    model_config = ConfigDict(from_attributes=True)

    id: int
    tenant_id: int
    report_code: str
    purchase_order_id: int
    purchase_order_item_id: int
    order_code: str
    material_id: int
    material_code: str
    material_name: str
    supplier_id: Optional[int] = None
    supplier_name: Optional[str] = None
    planned_arrival_date: date
    delay_reason: str
    estimated_arrival_date: date
    impact_description: Optional[str] = None
    impacted_assembly_summary: Optional[str] = None
    status: str
    review_status: str
    reviewer_id: Optional[int] = None
    reviewer_name: Optional[str] = None
    review_time: Optional[datetime] = None
    review_remarks: Optional[str] = None
    purchase_order_change_id: Optional[int] = None
    purchase_order_change_code: Optional[str] = None
    attachments: Optional[List[dict]] = None
    notes: Optional[str] = None
    created_by: Optional[int] = None
    created_by_name: Optional[str] = None
    updated_by: Optional[int] = None
    updated_by_name: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class PurchaseArrivalDelayReportListResponse(BaseModel):
    items: List[PurchaseArrivalDelayReportResponse]
    total: int
    skip: int
    limit: int


class PurchaseArrivalWarningListResponse(BaseModel):
    data: List[dict]
    total: int
    success: bool = True
    summary: dict

File: riveredge-frontend/src/apps/kuaizhizao/services/purchase-arrival.ts
import { apiRequest } from '../../../services/api';

export type PurchaseArrivalWarningLevel = 'normal' | 'imminent' | 'overdue';

export type PurchaseArrivalProcessingStatus =
  | 'unprocessed'
  | 'reported'
  | 'pending_review'
  | 'approved'
  | 'change_pending'
  | 'changed'
  | 'rejected';

export interface PurchaseArrivalWarningRow {
  id: number;
  purchase_order_id: number;
  purchase_order_item_id: number;
  order_code: string;
  supplier_name?: string;
  buyer_name?: string;
  material_code?: string;
  material_name?: string;
  required_date?: string;
  outstanding_quantity?: number;
  warning_level?: PurchaseArrivalWarningLevel;
  day_offset?: number;
  remaining_days?: number;
  overdue_days?: number;
  processing_status?: PurchaseArrivalProcessingStatus;
  impacted_assembly?: string;
  delay_report_id?: number;
  purchase_order_change_id?: number;
}

export interface PurchaseArrivalWarningListResult {
  data: PurchaseArrivalWarningRow[];
  total: number;
  success: boolean;
  summary: {
    normal: number;
    imminent: number;
    overdue: number;
    total_open_lines: number;
  };
}

export interface PurchaseArrivalDelayReport {
  id: number;
  report_code: string;
  purchase_order_id: number;
  purchase_order_item_id: number;
  order_code: string;
  material_code: string;
  material_name: string;
  planned_arrival_date: string;
  delay_reason: string;
  estimated_arrival_date: string;
  impact_description?: string;
  impacted_assembly_summary?: string;
  status: string;
  review_status: string;
  purchase_order_change_id?: number;
  purchase_order_change_code?: string;
}

export interface PurchaseArrivalWarningListParams {
  skip?: number;
  limit?: number;
  warning_level?: PurchaseArrivalWarningLevel;
  supplier_keyword?: string;
  order_code?: string;
  material_keyword?: string;
  processing_status?: PurchaseArrivalProcessingStatus;
}

export async function listPurchaseArrivalWarnings(
  params: PurchaseArrivalWarningListParams = {},
): Promise<PurchaseArrivalWarningListResult> {
  return apiRequest('/apps/kuaizhizao/purchase-arrival/warnings', { method: 'GET', params });
}

export async function createPurchaseArrivalDelayReport(data: {
  purchase_order_item_id: number;
  delay_reason: string;
  estimated_arrival_date: string;
  impact_description?: string;
  notes?: string;
}): Promise<PurchaseArrivalDelayReport> {
  return apiRequest('/apps/kuaizhizao/purchase-arrival/delay-reports', { method: 'POST', data });
}

export async function submitPurchaseArrivalDelayReport(id: number): Promise<PurchaseArrivalDelayReport> {
  return apiRequest(`/apps/kuaizhizao/purchase-arrival/delay-reports/${id}/submit`, { method: 'POST' });
}

export async function approvePurchaseArrivalDelayReport(
  id: number,
  approved: boolean,
  reviewRemarks?: string,
): Promise<PurchaseArrivalDelayReport> {
  return apiRequest(`/apps/kuaizhizao/purchase-arrival/delay-reports/${id}/approve`, {
    method: 'POST',
    data: { approved, review_remarks: reviewRemarks },
  });
}

export const PURCHASE_ARRIVAL_DELAY_REASONS = [
  { value: 'supplier_capacity', labelKey: 'app.kuaizhizao.purchaseArrival.delayReason.supplierCapacity' },
  { value: 'logistics', labelKey: 'app.kuaizhizao.purchaseArrival.delayReason.logistics' },
  { value: 'quality_rework', labelKey: 'app.kuaizhizao.purchaseArrival.delayReason.qualityRework' },
  { value: 'drawing_change', labelKey: 'app.kuaizhizao.purchaseArrival.delayReason.drawingChange' },
  { value: 'payment_delay', labelKey: 'app.kuaizhizao.purchaseArrival.delayReason.paymentDelay' },
  { value: 'other', labelKey: 'app.kuaizhizao.purchaseArrival.delayReason.other' },
] as const;
